File: solutions/go/raindrops/2/raindrops.go
package raindrops

import "strconv"

// Convert returns string based on given rules applied to number input as specified:
// if number has 3 as a factor, add 'Pling' to the result
// if number has 5 as a factor, add 'Plang' to the result
// if number has 7 as a factor, add 'Plong' to the result
// if number does not have any of 3, 5, or 7 as a factor, the result should be the digits of the number
func Convert(number int) string {
	result := maybeString("Pling", number % 3 == 0) + 
    	maybeString("Plang", number % 5 == 0) + 
    	maybeString("Plong", number % 7 == 0)

	if result == "" {
		return strconv.Itoa(number)
	}

	return result
}

func maybeString(str string, condition bool) string {
    if condition {
        return str
    }

    return ""
}

File: solutions/elixir/raindrops/2/lib/raindrops.ex
defmodule Raindrops do
  @doc """
  Returns a string based on raindrop factors.

  - If the number contains 3 as a prime factor, output 'Pling'.
  - If the number contains 5 as a prime factor, output 'Plang'.
  - If the number contains 7 as a prime factor, output 'Plong'.
  - If the number does not contain 3, 5, or 7 as a prime factor,
    just pass the number's digits straight through.
  """
  @word_conversion_map %{
    3 => "Pling",
    5 => "Plang",
    7 => "Plong",
  }

  @spec convert(pos_integer) :: String.t()
  def convert(number) do
    result = 3..7//2
    |> Enum.map(&(maybe_string(@word_conversion_map[&1], rem(number, &1) === 0)))
    |> Enum.join

    if result !== "", do: result, else: Integer.to_string(number)
  end

  defp maybe_string(_, false), do: ""
  defp maybe_string(string, _), do: string
end
